tests/connectors/wavefront_test.py

```diff
@@ -0,0 +1,459 @@
+            summarization="LAST",
+        )
+        assert metric.summarization.isalpha()
+
+    # Query
+    def test_query_required(self):
+        try:
+            WavefrontMetric(
+                name="throughput", unit=Unit.REQUESTS_PER_MINUTE, query=None
+            )
+        except ValidationError as error:
+            assert {
+                "loc": ("query",),
+                "msg": "none is not an allowed value",
+                "type": "type_error.none.not_allowed",
+            } in error.errors()
+
+
+class TestWavefrontConfiguration:
+    def test_url_required(self):
+        try:
+            WavefrontConfiguration(base_url=None)
+        except ValidationError as error:
+            assert {
+                "loc": ("base_url",),
+                "msg": "none is not an allowed value",
+                "type": "type_error.none.not_allowed",
+            } in error.errors()
+
+    def test_base_url_is_rstripped(self):
+        config = WavefrontConfiguration(
+            base_url="http://wavefront.com/some/path/", metrics=[]
+        )
+        assert config.base_url == "http://wavefront.com/some/path"
+
+    def test_supports_localhost_url(self):
+        config = WavefrontConfiguration(base_url="http://localhost:2878", metrics=[])
+        assert config.base_url == "http://localhost:2878"
+
+    def test_supports_cluster_url(self):
+        config = WavefrontConfiguration(
+            base_url="http://wavefront.com:2878", metrics=[]
+        )
+        assert config.base_url == "http://wavefront.com:2878"
+
+    def test_rejects_invalid_url(self):
+        try:
+            WavefrontConfiguration(base_url="gopher://this-is-invalid")
+        except ValidationError as error:
+            assert {
+                "loc": ("base_url",),
+                "msg": "URL scheme not permitted",
+                "type": "value_error.url.scheme",
+                "ctx": {
+                    "allowed_schemes": {
+                        "http",
+                        "https",
+                    },
+                },
+            } in error.errors()
+
+    def test_api_url(self):
+        config = WavefrontConfiguration(
+            base_url="http://wavefront.com:2878", metrics=[]
+        )
+        assert (
+            config.api_url == "http://wavefront.com:2878/api/v2/"
+        )
+
+    # Metrics
+    def test_metrics_required(self):
+        try:
+            WavefrontConfiguration(metrics=None)
+        except ValidationError as error:
+            assert {
+                "loc": ("metrics",),
+                "msg": "none is not an allowed value",
+                "type": "type_error.none.not_allowed",
+            } in error.errors()
+
+    # Generation
+    def test_generate_default_config(self):
+        config = WavefrontConfiguration.generate()
+        # Ugly patch for assertion until I understand why generate().yaml() metric order is being sorted alphabetically
```

Comment:
This is why hate winding up with so many branches in the air but I may have already sorted this out on another branch. I wound up switching to using [ruamel](https://yaml.readthedocs.io/en/latest/) for YAML serialization for some kind of issue with values gettiing mangled but I can't find the bits at the moment.

Ruamel is already a transient dependency through another library on tree, so you may want to try doing a `poetry add ruamel` and then updating `servo.config.AbstractBaseConfiguration.yaml` to pass it through ruamel instead of the stock library.

Ruamel supports round-tripping comments and maintaining ordering and some other stuff so I expect that we'll standardize on it anyway because it would be really nice to emit comments when you generate a config file.

Reply:
Ok so I was getting a bit of a headache trying to get ruamel to work. In the meantime I do now have the yaml generating properly with this one-liner:

 ```yaml.add_representer(dict, lambda self, data: yaml.representer.SafeRepresenter.represent_dict(self, data.items()))```

Reply:
You can pull this out. I just added the `sort_keys=False` argument on master (which I was already using in the CLI but missed in the model, derp)